File: Sources/SSH/Socket.swift
// FoxTerm | Socket.swift
// Copyright (c) 2025-2026 foxterm.app
// Created by foxterm@foxmail.com

import CSSH2
import Extension
import Foundation
import libetos
import Proxy
import Darwin

public extension SSH {
    /// 发起标准的 TCP 直接连接
    /// 使用 libetos 库进行非阻塞/带超时的 Socket 初始化
    /// - Returns: 是否连接成功（Bool）
    func connect() async -> Bool {
        // 异步调用底层 IO 执行连接操作
        fd = await io.Call { [self] in
            etos_socket_connect(host, port.int32, timeout.int32 * 1000)
        }
        // 校验连接状态，若失败则提取并保存错误信息
        guard isConnected else {
            error = socketLastStrError
            return false
        }
        return true
    }

    /// 通过代理服务器发起连接
    /// 支持 SOCKS5、HTTP 代理
    /// - Parameter proxy: 代理配置信息对象
    /// - Returns: 是否连接成功（Bool）
    func connect(proxy: ProxyConfiguration) async -> Bool {
        // 使用代理配置发起异步连接
        fd = await proxy.connect(host: host, port: port)
        // 校验连接状态，若失败则提取并保存错误信息
        guard isConnected else {
            error = socketLastStrError
            return false
        }
        return true
    }

    /// 获取当前连接的远程地址
    var remoteAddr: (host: String, port: Int)? {
        let buffer = Buffer<CChar>(0x40)
        var port: Int32 = 0
        guard etos_socket_get_peer_info(fd, buffer.buffer, buffer.count, &port) == 0 else {
            return nil
        }
        let host = buffer.buffer.string
        return (host, port.int)
    }

    /// 获取当前连接的本地地址
    var localAddr: (host: String, port: Int)? {
        let buffer = Buffer<CChar>(0x40)
        var port: Int32 = 0
        guard etos_socket_get_local_info(fd, buffer.buffer, buffer.count, &port) == 0 else {
            return nil
        }
        let host = buffer.buffer.string
        return (host, port.int)
    }

//    /// 内部数据发送方法
//    /// - Parameters:
//    ///   - fd: 套接字句柄
//    ///   - buffer: 待发送数据指针
//    ///   - length: 数据长度
//    ///   - flags: 系统 send 标志位
//    /// - Returns: 实际发送的字节数，负值代表错误
//    func send(fd: Int32, buffer: UnsafeRawPointer, length: ssize_t, flags: CInt) -> Int {
//        let size = libssh2_send(fd, buffer, length, flags)
//        if size < 0 {
//            return size
//        }
//        // 原子增加全局发送流量统计
//        sendSize.add(size)
//        return size
//    }
//
//    /// 内部数据接收方法
//    /// - Parameters:
//    ///   - fd: 套接字句柄
//    ///   - buffer: 接收缓冲区指针
//    ///   - length: 预期接收长度
//    ///   - flags: 系统 recv 标志位
//    /// - Returns: 实际接收的字节数，负值代表错误
//    func recv(fd: Int32, buffer: UnsafeMutableRawPointer, length: ssize_t, flags: CInt)
//        -> Int
//    {
//        let size = libssh2_recv(fd, buffer, length, flags)
//        if size < 0 {
//            return size
//        }
//        // 原子增加全局接收流量统计
//        recvSize.add(size)
//        return size
//    }

    /// 当前 Socket 的网络流量统计
    /// - Returns: 元组 (send: 已发送字节数, recv: 已接收字节数, rtt: 实时往返时间（微秒))
    var trafficStats: (send: UInt64, recv: UInt64, rtt: UInt32) {
        guard fd >= 0 else { return (0, 0, 0) }
        var stats = FdTrafficStats()
        guard etos_socket_get_traffic_stats(fd, &stats) == 0 else {
            return (0, 0, 0)
        }
        let tx = etos_stats_get_tx(&stats)
        let rx = etos_stats_get_rx(&stats)
        let rtt = etos_stats_get_rtt(&stats)
        return (tx, rx, rtt)
    }

    /// 检查底层 Socket 是否处于已连接状态
    var isConnected: Bool {
        etos_socket_is_connect(fd)
    }

    /// 获取底层 Socket 的最近一次错误码
    var socketLastError: Int32 {
        etos_socket_last_error()
    }

    /// 获取底层 Socket 的最近一次错误描述字符串
    var socketLastStrError: String {
        etos_socket_strerror(socketLastError).string
    }

    /// 等待套接字就绪（配合 libssh2 的非阻塞 IO 进行事件轮询）
    /// - Returns: 套接字是否正常/可继续操作
    func waitSocket() -> Bool {
        guard rawSession != nil, isConnected, fd >= 0 else {
            return false
        }

        let dir = libssh2_session_block_directions(rawSession)

        var pfd = Darwin.pollfd()
        pfd.fd = fd
        pfd.events = 0
        pfd.revents = 0

        if dir == 0 {
            pfd.events |= POLLIN.int16
        } else {
            if (dir & LIBSSH2_SESSION_BLOCK_INBOUND) != 0 {
                pfd.events |= POLLIN.int16
            }
            if (dir & LIBSSH2_SESSION_BLOCK_OUTBOUND) != 0 {
                pfd.events |= POLLOUT.int16
            }
        }

        // 调用 C 标准库 poll，超时时间为 10 毫秒
        let rc = Darwin.poll(&pfd, 1, 10)
        if rc < 0 {
            // poll 执行出错
            return false
        } else if rc == 0 {
            // 超时，套接字暂无数据就绪，返回 true 继续重试
            return true
        }

        // 检查是否有异常标志（错误、挂断、非法 fd）
        let revents = pfd.revents.int32
        if (revents & (POLLERR | POLLHUP | POLLNVAL)) != 0 {
            return false
        }

        return true
    }

    /// 执行 Socket 半关闭操作（关闭读或写通道）
    /// - Parameter how: 关闭类型（如只关发送或只关接收）
    func shutdown(_ how: Shout) {
        etos_socket_shutdown(fd, how.raw)
    }

    /// 彻底关闭并释放 Socket 资源
    func closeSocket() {
        etos_socket_close(fd)
        #if DEBUG
            print("♻️", "彻底关闭并释放 Socket 资源")
        #endif
    }
}

File: Sources/libetos/socket.c
#include "socket.h"

#include <arpa/inet.h>
#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <openssl/bio.h>
#include <openssl/buffer.h>
#include <openssl/evp.h>
#include <poll.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>

/* ------------------------------------------------------------
   内部辅助函数
   ------------------------------------------------------------ */
static void set_nosigpipe(int fd) {
#ifdef SO_NOSIGPIPE
  int optval = 1;
  setsockopt(fd, SOL_SOCKET, SO_NOSIGPIPE, &optval, sizeof(optval));
#else
  (void)fd;
#endif
}

static bool recv_exact(int fd, void *buf, size_t len, int timeout_ms) {
  size_t total_read = 0;
  char *ptr = (char *)buf;

  while (total_read < len) {
    ssize_t rc = etos_socket_recv_timeout(fd, ptr + total_read, len - total_read, 0, timeout_ms);
    if (rc <= 0) {
      return false;
    }
    total_read += rc;
  }
  return true;
}

static int connect_with_timeout(int fd, const struct sockaddr *addr, socklen_t addrlen, int timeout_ms) {
  if (timeout_ms <= 0) {
    return connect(fd, addr, addrlen);
  }

  if (etos_socket_set_blocking(fd, false) != 0) {
    return -1;
  }

  int ret = connect(fd, addr, addrlen);
  if (ret < 0 && errno != EINPROGRESS) {
    etos_socket_set_blocking(fd, true);
    return -1;
  }

  if (ret == 0) {
    etos_socket_set_blocking(fd, true);
    return 0;
  }

  struct pollfd pfd;
  pfd.fd = fd;
  pfd.events = POLLOUT | POLLIN;
  pfd.revents = 0;

  ret = poll(&pfd, 1, timeout_ms);
  if (ret <= 0) {
    etos_socket_set_blocking(fd, true);
    return -1;
  }

  int error = 0;
  socklen_t len = sizeof(error);
  if (getsockopt(fd, SOL_SOCKET, SO_ERROR, &error, &len) < 0 || error != 0) {
    if (error != 0)
      errno = error;
    etos_socket_set_blocking(fd, true);
    return -1;
  }

  etos_socket_set_blocking(fd, true);
  return 0;
}

static int extract_sockaddr_info(const struct sockaddr_storage *addr, char *ip_buf, size_t ip_buf_len, int *port) {
  if (addr->ss_family == AF_INET) {
    struct sockaddr_in *s = (struct sockaddr_in *)addr;
    *port = ntohs(s->sin_port);
    if (inet_ntop(AF_INET, &s->sin_addr, ip_buf, ip_buf_len) == NULL) {
      return -1;
    }
  } else if (addr->ss_family == AF_INET6) {
    struct sockaddr_in6 *s = (struct sockaddr_in6 *)addr;
    *port = ntohs(s->sin6_port);
    if (inet_ntop(AF_INET6, &s->sin6_addr, ip_buf, ip_buf_len) == NULL) {
      return -1;
    }
  } else {
    return -1;
  }
  return 0;
}

char *etos_base64_encode(const char *input) {
  if (!input)
    return NULL;
  BIO *b64 = BIO_new(BIO_f_base64());
  BIO *mem = BIO_new(BIO_s_mem());
  BIO_push(b64, mem);
  BIO_set_flags(b64, BIO_FLAGS_BASE64_NO_NL);
  BIO_write(b64, input, (int)strlen(input));
  BIO_flush(b64);
  BUF_MEM *ptr;
  BIO_get_mem_ptr(mem, &ptr);
  char *out = (char *)malloc(ptr->length + 1);
  if (out) {
    memcpy(out, ptr->data, ptr->length);
    out[ptr->length] = '\0';
  }
  BIO_free_all(b64);
  return out;
}

static bool handshake_http_proxy(int fd, const char *target_host, int target_port, const char *user, const char *password, int timeout_ms) {
  if (!target_host || target_port <= 0 || target_port > 65535) {
    return false;
  }

  char req[1024];
  int len = 0;

  char auth_header[512] = "";
  if (user && password && user[0] != '\0') {
    char auth_raw[256];
    snprintf(auth_raw, sizeof(auth_raw), "%s:%s", user, password);
    char *auth_b64 = etos_base64_encode(auth_raw);
    if (auth_b64) {
      snprintf(auth_header, sizeof(auth_header), "Proxy-Authorization: Basic %s\r\n", auth_b64);
      free(auth_b64);
    }
  }

  struct in6_addr dummy_v6;
  if (inet_pton(AF_INET6, target_host, &dummy_v6) == 1) {
    len = snprintf(req, sizeof(req),
                   "CONNECT [%s]:%d HTTP/1.1\r\n"
                   "Host: [%s]:%d\r\n"
                   "%s\r\n",
                   target_host, target_port, target_host, target_port, auth_header);
  } else {
    len = snprintf(req, sizeof(req),
                   "CONNECT %s:%d HTTP/1.1\r\n"
                   "Host: %s:%d\r\n"
                   "%s\r\n",
                   target_host, target_port, target_host, target_port, auth_header);
  }

  if (len < 0 || len >= (int)sizeof(req) || etos_socket_send_timeout(fd, req, len, 0, timeout_ms) <= 0) {
    return false;
  }

  char resp[2048];
  size_t resp_len = 0;
  bool header_complete = false;

  while (resp_len < sizeof(resp) - 1) {
    ssize_t rc = etos_socket_recv_timeout(fd, resp + resp_len, sizeof(resp) - 1 - resp_len, 0, timeout_ms);
    if (rc <= 0)
      break;
    resp_len += rc;
    resp[resp_len] = '\0';

    if (strstr(resp, "\r\n\r\n") != NULL) {
      header_complete = true;
      break;
    }
  }

  if (!header_complete) {
    return false;
  }

  if (strncasecmp(resp, "HTTP/1.0 200", 12) == 0 || strncasecmp(resp, "HTTP/1.1 200", 12) == 0) {
    return true;
  }

  return false;
}

static bool handshake_socks5_proxy(int fd, const char *target_host, int target_port, const char *user, const char *password, int timeout_ms) {
  if (!target_host || target_port <= 0 || target_port > 65535) {
    return false;
  }

  unsigned char auth_req[3];
  auth_req[0] = 0x05;
  auth_req[1] = 0x01;
  auth_req[2] = (user && password && user[0] != '\0') ? 0x02 : 0x00;

  if (etos_socket_send_timeout(fd, (char *)auth_req, 3, 0, timeout_ms) <= 0) {
    return false;
  }

  unsigned char auth_resp[2] = {0};
  if (!recv_exact(fd, auth_resp, 2, timeout_ms) || auth_resp[0] != 0x05) {
    return false;
  }

  if (auth_resp[1] == 0x02) {
    if (!user || !password)
      return false;

    size_t ulen = strlen(user);
    size_t plen = strlen(password);
    if (ulen > 255 || plen > 255)
      return false;

    unsigned char pass_req[512];
    size_t pass_len = 0;
    pass_req[pass_len++] = 0x01;
    pass_req[pass_len++] = (unsigned char)ulen;
    memcpy(&pass_req[pass_len], user, ulen);
    pass_len += ulen;
    pass_req[pass_len++] = (unsigned char)plen;
    memcpy(&pass_req[pass_len], password, plen);
    pass_len += plen;

    if (etos_socket_send_timeout(fd, (char *)pass_req, pass_len, 0, timeout_ms) <= 0) {
      return false;
    }

    unsigned char pass_resp[2] = {0};
    if (!recv_exact(fd, pass_resp, 2, timeout_ms) || pass_resp[1] != 0x00) {
      return false;
    }
  } else if (auth_resp[1] != 0x00) {
    return false;
  }

  unsigned char conn_req[300];
  size_t p = 0;

  conn_req[p++] = 0x05;
  conn_req[p++] = 0x01;
  conn_req[p++] = 0x00;

  struct in_addr addr4;
  struct in6_addr addr6;

  if (inet_pton(AF_INET, target_host, &addr4) == 1) {
    conn_req[p++] = 0x01;
    memcpy(&conn_req[p], &addr4, 4);
    p += 4;
  } else if (inet_pton(AF_INET6, target_host, &addr6) == 1) {
    conn_req[p++] = 0x04;
    memcpy(&conn_req[p], &addr6, 16);
    p += 16;
  } else {
    size_t target_len = strlen(target_host);
    if (target_len > 255)
      return false;
    conn_req[p++] = 0x03;
    conn_req[p++] = (unsigned char)target_len;
    memcpy(&conn_req[p], target_host, target_len);
    p += target_len;
  }

  unsigned short net_port = htons((unsigned short)target_port);
  memcpy(&conn_req[p], &net_port, 2);
  p += 2;

  if (etos_socket_send_timeout(fd, (char *)conn_req, p, 0, timeout_ms) <= 0) {
    return false;
  }

  unsigned char head[4];
  if (!recv_exact(fd, head, 4, timeout_ms)) {
    return false;
  }

  if (head[0] != 0x05 || head[1] != 0x00) {
    return false;
  }

  size_t skip_bytes = 0;
  if (head[3] == 0x01) {
    skip_bytes = 4 + 2;
  } else if (head[3] == 0x04) {
    skip_bytes = 16 + 2;
  } else if (head[3] == 0x03) {
    unsigned char dlen = 0;
    if (!recv_exact(fd, &dlen, 1, timeout_ms))
      return false;
    skip_bytes = (size_t)dlen + 2;
  } else {
    return false;
  }

  unsigned char dummy[260];
  if (!recv_exact(fd, dummy, skip_bytes, timeout_ms)) {
    return false;
  }

  return true;
}

/* ------------------------------------------------------------
   外部接口实现
   ------------------------------------------------------------ */
int etos_socket_get_traffic_stats(int fd, FdTrafficStats *stats) {
  if (fd < 0 || !stats) {
    return -1;
  }

  struct tcp_connection_info info;
  socklen_t len = sizeof(info);

  if (getsockopt(fd, IPPROTO_TCP, TCP_CONNECTION_INFO, &info, &len) == 0) {
    atomic_store(&stats->rx_bytes, info.tcpi_rxbytes);
    atomic_store(&stats->tx_bytes, info.tcpi_txbytes);
    atomic_store(&stats->rtt_us, info.tcpi_rttcur);
    return 0;
  }

  return -1;
}

u_int64_t etos_stats_get_rx(const FdTrafficStats *stats) {
  if (!stats)
    return 0;
  return atomic_load(&stats->rx_bytes);
}

u_int64_t etos_stats_get_tx(const FdTrafficStats *stats) {
  if (!stats)
    return 0;
  return atomic_load(&stats->tx_bytes);
}

u_int32_t etos_stats_get_rtt(const FdTrafficStats *stats) {
  if (!stats)
    return 0;
  return atomic_load(&stats->rtt_us);
}

int etos_socket_set_keepalive(int fd, bool enable, int idle_sec, int interval_sec, int count) {
  int optval = enable ? 1 : 0;
  if (setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, &optval, sizeof(optval)) < 0) {
    return -1;
  }

  if (enable) {
    if (idle_sec > 0) {
      setsockopt(fd, IPPROTO_TCP, TCP_KEEPALIVE, &idle_sec, sizeof(idle_sec));
    }
    if (interval_sec > 0) {
      setsockopt(fd, IPPROTO_TCP, TCP_KEEPINTVL, &interval_sec, sizeof(interval_sec));
    }
    if (count > 0) {
      setsockopt(fd, IPPROTO_TCP, TCP_KEEPCNT, &count, sizeof(count));
    }
  }
  return 0;
}

int etos_socket_set_nodelay(int fd, bool enable) {
  int optval = enable ? 1 : 0;
  return setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &optval, sizeof(optval));
}

int etos_socket_connect(const char *host, int port, int timeout_ms) {
  if (!host || port <= 0 || port > 65535)
    return ETOS_INVALID_SOCKET;

  // 清洗主机名，剥离方括号
  char clean_host[256];
  size_t host_len = strlen(host);
  if (host[0] == '[' && host[host_len - 1] == ']' && host_len < sizeof(clean_host) + 2) {
    strncpy(clean_host, host + 1, host_len - 2);
    clean_host[host_len - 2] = '\0';
  } else {
    strncpy(clean_host, host, sizeof(clean_host) - 1);
    clean_host[sizeof(clean_host) - 1] = '\0';
  }

  char port_str[16];
  snprintf(port_str, sizeof(port_str), "%d", port);

  struct addrinfo hints, *res = NULL, *rp = NULL;
  memset(&hints, 0, sizeof(hints));
  hints.ai_family = AF_UNSPEC;
  hints.ai_socktype = SOCK_STREAM;

  struct in6_addr dummy_v6;
  struct in_addr dummy_v4;
  if (inet_pton(AF_INET6, clean_host, &dummy_v6) == 1 || inet_pton(AF_INET, clean_host, &dummy_v4) == 1) {
    hints.ai_flags |= AI_NUMERICHOST;
  }

  if (getaddrinfo(clean_host, port_str, &hints, &res) != 0) {
    return ETOS_INVALID_SOCKET;
  }

  int fd = ETOS_INVALID_SOCKET;
  for (rp = res; rp != NULL; rp = rp->ai_next) {
    fd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
    if (fd < 0)
      continue;

    set_nosigpipe(fd);

    if (connect_with_timeout(fd, rp->ai_addr, rp->ai_addrlen, timeout_ms) == 0) {
      break;
    }

    close(fd);
    fd = ETOS_INVALID_SOCKET;
  }

  if (res) {
    freeaddrinfo(res);
  }

  if (fd != ETOS_INVALID_SOCKET) {
    etos_socket_set_nodelay(fd, true);
    etos_socket_set_keepalive(fd, true, 5, 5, 10);
  }

  return fd;
}

int etos_socket_connect_proxy(int type, const char *proxy_host, int proxy_port, int timeout_ms, const char *target_host, int target_port, const char *user, const char *password) {
  if (type == ETOS_PROXY_NONE) {
    return etos_socket_connect(target_host, target_port, timeout_ms);
  }

  int fd = etos_socket_connect(proxy_host, proxy_port, timeout_ms);
  if (fd < 0)
    return ETOS_INVALID_SOCKET;

  bool ok = false;
  if (type == ETOS_PROXY_HTTP) {
    ok = handshake_http_proxy(fd, target_host, target_port, user, password, timeout_ms);
  } else if (type == ETOS_PROXY_SOCKS5) {
    ok = handshake_socks5_proxy(fd, target_host, target_port, user, password, timeout_ms);
  }

  if (!ok) {
    etos_socket_close(fd);
    return ETOS_INVALID_SOCKET;
  }

  return fd;
}

/* 全量 poll 姿势实现发送超时 */
ssize_t etos_socket_send_timeout(int fd, const char *buf, size_t len, int flags, int timeout_ms) {
  if (timeout_ms > 0) {
    struct pollfd pfd;
    pfd.fd = fd;
    pfd.events = POLLOUT;
    pfd.revents = 0;

    int ret = poll(&pfd, 1, timeout_ms);
    if (ret <= 0)
      return -ETIMEDOUT;
  }
  return etos_socket_send(fd, buf, len, flags);
}

/* 全量 poll 姿势实现接收超时 */
ssize_t etos_socket_recv_timeout(int fd, char *buf, size_t len, int flags, int timeout_ms) {
  if (timeout_ms > 0) {
    struct pollfd pfd;
    pfd.fd = fd;
    pfd.events = POLLIN;
    pfd.revents = 0;

    int ret = poll(&pfd, 1, timeout_ms);
    if (ret <= 0)
      return -ETIMEDOUT;
  }
  return etos_socket_recv(fd, buf, len, flags);
}

ssize_t etos_socket_send(int fd, const char *buf, size_t len, int flags) { return send(fd, buf, len, flags); }

ssize_t etos_socket_recv(int fd, char *buf, size_t len, int flags) { return recv(fd, buf, len, flags); }

int etos_socket_shutdown(int fd, int how) { return shutdown(fd, how); }

void etos_socket_close(int fd) {
  if (fd >= 0) {
    close(fd);
  }
}

int etos_socket_set_blocking(int fd, bool blocking) {
  int flags = fcntl(fd, F_GETFL, 0);
  if (flags < 0)
    return -1;

  if (blocking) {
    flags &= ~O_NONBLOCK;
  } else {
    flags |= O_NONBLOCK;
  }
  return fcntl(fd, F_SETFL, flags);
}

bool etos_socket_is_connect(int fd) {
  if (fd < 0)
    return false;

  char buf;
  ssize_t res = recv(fd, &buf, 1, MSG_PEEK | MSG_DONTWAIT);
  if (res == 0) {
    return false;
  }
  if (res < 0) {
    if (errno == EAGAIN || errno == EWOULDBLOCK) {
      return true;
    }
    return false;
  }
  return true;
}

int etos_socket_last_error(void) { return errno; }

const char *etos_socket_strerror(int errnum) { return strerror(errnum); }

int etos_socket_get_peer_info(int fd, char *ip_buf, size_t ip_buf_len, int *port) {
  if (fd < 0 || !ip_buf || ip_buf_len == 0 || !port) {
    return -1;
  }

  struct sockaddr_storage addr;
  socklen_t addr_len = sizeof(addr);

  if (getpeername(fd, (struct sockaddr *)&addr, &addr_len) < 0) {
    return -1;
  }

  return extract_sockaddr_info(&addr, ip_buf, ip_buf_len, port);
}

int etos_socket_get_local_info(int fd, char *ip_buf, size_t ip_buf_len, int *port) {
  if (fd < 0 || !ip_buf || ip_buf_len == 0 || !port) {
    return -1;
  }

  struct sockaddr_storage addr;
  socklen_t addr_len = sizeof(addr);

  if (getsockname(fd, (struct sockaddr *)&addr, &addr_len) < 0) {
    return -1;
  }

  return extract_sockaddr_info(&addr, ip_buf, ip_buf_len, port);
}
